src/librustc/ty/print/pretty.rs

```diff
@@ -969,69 +987,131 @@ pub trait PrettyPrinter<'tcx>:
+            // Char
+            (Scalar::Raw { data, .. }, ty::Char) =>
                 p!(write("{:?}", ::std::char::from_u32(data as u32).unwrap())),
-            (ConstValue::Scalar(_), ty::RawPtr(_)) => p!(write("{{pointer}}")),
-            (ConstValue::Scalar(Scalar::Ptr(ptr)), ty::FnPtr(_)) => {
+            // References and pointers
+            (Scalar::Raw { data: 0, .. }, ty::RawPtr(_)) => p!(write("{{null pointer}}")),
+            // This is UB, but we still print it
+            (Scalar::Raw { data: 0, ..}, ty::Ref(_, ty, _))
+                => p!(write("{{null reference to "), print(ty), write("}}")),
+            (Scalar::Raw { data, .. }, ty::Ref(..)) |
+            (Scalar::Raw { data, .. }, ty::RawPtr(_)) => {
+                let pointer_width = self.tcx().data_layout.pointer_size.bytes();
+                p!(write("0x{:01$x}", data, pointer_width as usize * 2))
+            },
+            (Scalar::Ptr(p), ty::Ref(..)) |
+            (Scalar::Ptr(p), ty::RawPtr(_)) => {
+                self = self.pretty_print_const_pointer(p)?
+            },
+            (Scalar::Ptr(ptr), ty::FnPtr(_)) => {
                 let instance = {
                     let alloc_map = self.tcx().alloc_map.lock();
                     alloc_map.unwrap_fn(ptr.alloc_id)
                 };
                 p!(print_value_path(instance.def_id(), instance.substs));
             },
-            _ => {
-                let printed = if let ty::Ref(_, ref_ty, _) = ty.kind {
-                    let byte_str = match (ct, &ref_ty.kind) {
-                        (ConstValue::Scalar(Scalar::Ptr(ptr)), ty::Array(t, n)) if *t == u8 => {
-                            let n = n.eval_usize(self.tcx(), ty::ParamEnv::empty());
-                            Some(self.tcx()
-                                .alloc_map.lock()
-                                .unwrap_memory(ptr.alloc_id)
-                                .get_bytes(&self.tcx(), ptr, Size::from_bytes(n)).unwrap())
-                        },
-                        (ConstValue::Slice { data, start, end }, ty::Slice(t)) if *t == u8 => {
-                            // The `inspect` here is okay since we checked the bounds, and there are
-                            // no relocations (we have an active slice reference here). We don't use
-                            // this result to affect interpreter execution.
-                            Some(data.inspect_with_undef_and_ptr_outside_interpreter(start..end))
-                        },
-                        _ => None,
-                    };
+            // For zsts just print their type as their value gives no extra information
+            (Scalar::Raw { size: 0, .. }, _) => p!(print(ty)),
+            // Nontrivial types with scalar bit representation
+            (Scalar::Raw { data, size }, _) => p!(
+                write("{{0x{:01$x} : ", data, size as usize * 2),
+                print(ty),
```

Comment:
this is where I am getting the `Option::<bool>` instead of `Option<bool>`. If I move back to printing `ty` via `write` and `{}` I get back the `Option<bool>`, should I do that?